### PR TITLE
Add mutex feature

### LINKED ISSUE #120: Guard libcryptsetup calls with a mutex internally
@mulkieran Currently, we're doing this in stratisd but we may have more control over this in the `Drop` implementations for example and protect other users of libcryptsetup-rs if we do this in the library instead. It would also simplify the code for stratisd. What do you think? Do you want me to investigate this further?

### PR DESCRIPTION
**Pull request by jbaublitz**
Closes #120

### CODE REVIEW

**Comment by jbaublitz** [comment]
I'm currently testing against `make test-real` on stratisd to make sure that this doesn't introduce any deadlocks. So far, so good.

**Comment by jbaublitz** [comment]
No deadlocks and the tests all passed for stratisd. I think this is ready for review.

File: src/macros.rs [hunk 1]
```diff
@@ -2,6 +2,31 @@
...
+                    Locking failed with error: {}; see Rust's documentation \
+                    on Mutex poisoning here to decide how to proceed: \
+                    https://doc.rust-lang.org/std/sync/struct.Mutex.html#errors",
+                    e,
+                );
+                None
+            }
+        };
+
+        $libcryptsetup_call
```
Comment (mulkieran) [inline src/macros.rs:26]: I believe that every invocation that requires the ```mutex!``` must also be necessarily ```unsafe```. Is there any benefit to including the ```unsafe``` in this macro definition?

**Review by mulkieran — CHANGES_REQUESTED** [review]
This all makes sense to me. I found one typo, and I have a question.

File: src/macros.rs (hunk 1 above)
Comment (jbaublitz) [reply src/macros.rs:26]: Let me think about this. I can see a benefit in terms of code readability, but I'm still thinking through if it'll be confusing obscuring that from view. I know in other places, we've avoided returning errors in macros and have opted instead to put the `?` operator outside of the macro. Do you think the same thing applies here?

File: src/macros.rs [hunk 2]
```diff
@@ -2,6 +2,31 @@
...
+/// a no-op.
+macro_rules! mutex {
+    ( $libcryptsetup_call:expr ) => {{
+        #[cfg(feature = "mutex")]
+        #[allow(unused_variables)]
+        let lock = match $crate::MUTEX.lock() {
+            Ok(l) => Some(l),
+            Err(e) => {
+                log::warn!(
+                    "The synchronization mutex can not longer be locked! \
```
[the remark this answers is not in the document]
Comment (jbaublitz) [reply src/macros.rs:16]: Ah yes. I'll fix this.

[the remark this answers is not in the document]
Comment (jbaublitz) [reply src/macros.rs]: Resolved.

File: src/macros.rs (hunk 1 above)
Comment (mulkieran) [reply src/macros.rs:26]: I'm not sure...one thing I thought was that if the unsafe call around the libcryptsetup invocation was mandatory (if missing a compiler error results), including the unsafe in the macro would ensure that the mutex was always added in the right place in new code.

Comment (jbaublitz) [reply src/macros.rs:26]: That seems like a good reason to me. I'm willing to update the PR with that.

Comment (jbaublitz) [reply src/macros.rs:26]: Resolved.

**Review by mulkieran — CHANGES_REQUESTED** [review]
Looks good! Please close and open final PR.

**Comment by jbaublitz** [comment]
@mulkieran The Cargo dependencies have been bumped and I've tested with mock. It passes.

**Pull request closed without merging**